packages/server-wallet/src/chain-service/chain-service.ts

```diff
@@ -411,15 +411,103 @@ export class ChainService implements ChainServiceInterface {
 
   private async getInitialHoldings(
     contract: Contract,
-    channelId: string
-  ): Promise<HoldingUpdatedArg> {
-    const holding = BN.from(await contract.holdings(channelId));
-
-    return {
-      channelId,
-      assetHolderAddress: makeAddress(contract.address),
-      amount: BN.from(holding),
-    };
+    channelId: string,
+    eventTracker: EventTracker
+  ): Promise<void> {
+    const currentBlock = await this.provider.getBlockNumber();
+    const confirmedBlock = currentBlock - this.blockConfirmations + 1;
+    const currentHolding = BN.from(await contract.holdings(channelId));
+    let confirmedHolding = BN.from(0);
+    try {
+      // https://docs.ethers.io/v5/api/contract/contract/#Contract--metaclass
+      // provider can lose track of the latest block, force it to reload
+      const overrides = {
+        blockTag: confirmedBlock,
+      };
+
+      confirmedHolding = BN.from(await contract.holdings(channelId, overrides));
+      this.logger.debug(
+        `Successfully red confirmedHoldings (${confirmedHolding}), from block ${confirmedBlock}.`
+      );
+    } catch (e) {
+      // We should have e.code="CALL_EXCEPTION", but gaven currentHolding query succeeded,
+      // the cause of exception is obvious
+      this.logger.debug(
+        `Error caught while trying to query confirmed holding in block ${confirmedBlock},` +
+          `this is likely due to the channel is new, safe to ignore, error is ${JSON.stringify(e)}`
+      );
+    }
+
+    this.logger.debug(
+      `getConfirmedHoldings: from block ${confirmedBlock}, confirmed holding ${confirmedHolding}, current holding ${currentHolding}.`
+    );
+
+    // Report confirmed holding immediately
+    eventTracker.holdingUpdated(
+      {
+        channelId: channelId,
+        assetHolderAddress: makeAddress(contract.address),
+        amount: confirmedHolding,
+      },
+      confirmedBlock
+    );
+
+    // We're unsure if the same events are also played by contract observer callback,
+    // but need to play it regardless, so subscribers won't miss anything.
+    // See EventTracker documentation
+    for (const ethersEvent of (await contract.queryFilter({}, confirmedBlock)).sort(
+      e => e.blockNumber
+    )) {
+      await this.waitForConfirmations(ethersEvent);
```

Comment:
It seems like this is the same logic that is applied for events that are generated by contract listeners. Can it be factored out?